### ISSUE TITLE
Seeking in a film is not working in reverse mode

### ISSUE THREAD

**Issue by pletexier** [euzu/m3u-filter#93]
Hi @euzu,

Seeking in a film is not working in Reverse proxy mode.
I think this is the same problem as issue #82 .

Regards,

Patrice

**Comment by euzu** [comment]
@pletexier can you fünd out what url is called?
it could be missing api implementation.

**Comment by pletexier** [comment]
I think the "range" header is ignored when loading the video file.

**Comment by euzu** [comment]
As you can see in the code the headers are forwarded: [stream proxy](https://github.com/euzu/m3u-filter/blob/02b0171c9fc0797c0e8887e0f3d42eca154ad4dc/src/api/api_utils.rs#L57)

**Comment by euzu** [comment]
@pletexier i have tested it and i could seek forward and backward. There must be something different with your provider or client. We need to figure out what it is.

**Comment by euzu** [comment]
@pletexier do you have any updates regarding the problem?

**Comment by euzu** [comment]
@pletexier with the latest version some users reported that this is working for them. Can you confirm?

**Comment by euzu** [comment]
closed due inactivity

**Comment by jojo141185** [comment]
I am experiencing the same problem with the m3u-filter proxy. Depending on the player, the stream starts from the beginning or continues at the same point after a short pause. 
In comparison with [iptv-proxy](https://github.com/pierre-emmanuelJ/iptv-proxy), seeking and skipping with the same stream works without any problems.

This is what I see repeatedly in the debug log of m3u-filter after trying to seek:

_Player: IPTVNator_
```
[2025-01-04T18:32:08Z DEBUG m3u_filter::api::xtream_api] Streaming stream request from http://myXtream.url/series/myUser/myPass/1753625.mkv
[2025-01-04T18:32:08Z DEBUG m3u_filter::api::api_utils] Try to open stream http://myXtream.url/series/myUser/myPass/1753625.mkv
[2025-01-04T18:32:08Z DEBUG m3u_filter::utils::request_utils] Ignoring request header host=192.168.2.10:8901
[2025-01-04T18:32:08Z DEBUG m3u_filter::utils::request_utils] Request headers {"accept-language": "de", "accept-encoding": "identity;q=1, *;q=0", "range": "bytes=0-", "user-agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) iptvnator/0.16.0 Chrome/118.0.5993.144 Electron/27.1.0 Safari/537.36", "accept": "*/*", "connection": "keep-alive"}
```

_Player: VLC_
```
[2025-01-04T18:34:18Z DEBUG m3u_filter::api::api_utils] Try to open stream http://myXtream.url/series/myUser/myPass/1753625.mkv
[2025-01-04T18:34:18Z DEBUG m3u_filter::utils::request_utils] Ignoring request header host=192.168.2.10:8901
[2025-01-04T18:34:18Z DEBUG m3u_filter::utils::request_utils] Request headers {"accept-language": "de", "user-agent": "VLC/3.0.20 LibVLC/3.0.20", "range": "bytes=0-", "accept": "*/*"}
[2025-01-04T18:44:12Z DEBUG m3u_filter::api::xtream_api] Streaming stream request from http://myXtream.url/series/myUser/myPass/1753625.mkv
[2025-01-04T18:44:12Z DEBUG m3u_filter::api::api_utils] Try to open stream http://myXtream.url/series/myUser/myPass/1753625.mkv
```

Example stream / codec tested:
![grafik](https://github.com/user-attachments/assets/2284de30-6418-49ad-ac34-c76ed6e4ad84)

Can we reopen the issue if necessary or is there a solution for this by now?
Is it perhaps possible to test the behavior using various freely available sources (e.g. m3u file)?

**Comment by euzu** [comment]
Yes we can reopen this issue. I will try to figure out why it is not working anymore. I know it worked.

**Comment by euzu** [comment]
@pletexier and  @jojo141185  i have pushed changes to the branch [reconnect_stream](https://github.com/euzu/m3u-filter/tree/feature/reconnect_stream)

which includes following changes for reverse proxy mode 
- stream reconnect and buffering
- Seeking

It would be nice if you can test it

**Comment by jojo141185** [comment]
@euzu Thank you euzu! After a short test, unfortunately nothing has changed for me with the latest adjustments. For example, when I try to jump to a certain point in the video, the stream always starts again from the beginning. Used VLC for testing.

**Comment by euzu** [comment]
I have done some tests and live tv is working with all clients,  but vod is working only on some clients. I will try to figure out the reason.

**Comment by jojo141185** [comment]
@euzu, I briefly tested your new approach with VLC. The behavior is better now - at least you can jump to a certain point in the video without the stream starting directly from the beginning. 
However, the further I jump to the end of the video, the longer it takes for playback to continue from that point.  It seems that the proxy retrieves the entire stream sequentially from the beginning to the desired position instead of retrieving the data directly at the desired offset. This can also be easily detected via the data traffic (e.g. with nethogs). 

Maybe we could try a simpler approach (without buffering) first. With the simple copy implementation, as implemented here in [iptv-proxy](https://github.com/pierre-emmanuelJ/iptv-proxy/blob/6561cba1e60cfd415fb6555b66524ba422a7aa09/pkg/server/handlers.go#L65), seeking and jumping to different places in the video worked quickly and without problems for me.

**Comment by euzu** [comment]
@jojo141185 the seeking works with range headers. each time you go forward or backward the client sends a new request with range. m3u filter forwards this request.

**Comment by euzu** [comment]
@jojo141185 the implementation is in src/api/model/buffered_stream.rs. You can see there how it is implemented. improvment ideas are welcome.

**Comment by euzu** [comment]
@jojo141185 i have pushed some changes because some clients do not work if you dont send 206 for partial content.

I have tested now seeking  with mobile phone, tv,  docker android  on linux,  hypnotix on linux.  It is working.

VLC  has some problems, i get 

```
[00007b448c048f00] main decoder error: Timestamp conversion failed for 43627001: no reference clock
[00007b448c048f00] main decoder error: Could not convert timestamp 0 for FFmpeg
[00007b448c048f00] main decoder error: Timestamp conversion failed for 55097001: no reference clock
```

And stops the seeking. I dont know why this happens.

**Comment by jojo141185** [comment]
@euzu Thank you for trying again!
The behavior on my side is identical to the [last attempt](https://github.com/euzu/m3u-filter/issues/93#issuecomment-2583974300)
I have looked at the error messages in VLC. The following errors don't seem to be a problem, I also get them when I connect directly or use iptv-proxy, where seeking works flawlessly and fast.
```
main error: Timestamp conversion failed for 4543001: no reference clock
main error: Could not convert timestamp 0 for FFmpeg
```
In conjunction with m3u-filter, however, I get an additional error when starting the stream, which may be the cause of the problem:
```
mkv error: cannot find any cluster or chapter, damaged file ?
....
```
This error does not show up when I connect directly or via iptv-proxy!

After seeking I get the following output in the debug log of m3u-filter:
```
[2025-01-11T16:19:49Z DEBUG m3u_filter::api::xtream_api] Streaming stream request from http://***/movie/***/1760331.mkv
[2025-01-11T16:19:49Z DEBUG m3u_filter::api::model::buffered_stream] Stream requested with headers: [(“accept”, “*/*”), (“range”, “bytes=938828118-”), (“host”, “192.168.2.10:8901”), (“user-agent”, “VLC/3.0.20 LibVLC/3.0.20”), (“accept-language”, “en”)]
[2025-01-11T16:19:50Z DEBUG m3u_filter::api::model::buffered_stream] Opening stream http://***/movie/***/1760331.mkv with status 206, headers [("content-type", "video/x-matroska"), ("content-length", "5009878879"), ("connection", "keep-alive"), ("accept-ranges", "0-5948706997"), ("content-range", "bytes 938828118-5948706996/5948706997"), ("vary", "accept-encoding"), ("date", "Sat, 11 Jan 2025 16:19:50 +0000")]
[2025-01-11T16:19:50Z DEBUG m3u_filter::api::model::buffered_stream] Client has disconnected from stream http://***/movie/***/1760331.mkv
[2025-01-11T16:19:51Z DEBUG m3u_filter::api::model::buffered_stream] Streaming stopped and no reconnect  for http://***/movie/***/1760331.mkv
```
Also the message “Client has disconnected from stream” although the player is still connected seems strange.

**Comment by euzu** [comment]
@jojo141185 The client initiates a new request for each seek operation. This can involve opening a new connection or reusing the existing connection if Connection: keep-alive is supported between the client and server.

When the connection is closed, the client calculates the byte position for the new seek and sends a range request to the server.

Yes, every seek operation results in a new request.

**Comment by euzu** [comment]
@jojo141185 how do you use vlc for xt api ?  or m3u playlist, when i enter the url it plays the first stream, i cant choose any channel?   I have tested it with the `https://files.catbox.moe/ach5l9.m3u` playlist and seeking works. But i cant find a source which has chapters like in your case.

**Comment by jojo141185** [comment]
@euzu You can open the playlist in VLC and see the different links from the m3u file in the playlist window:
1. Open VLC Media Player.
2. Click View in the menu bar.
3. Select Playlist (or press Ctrl + L)
The last video should have chapters. 
But seeking is working with these files included in the test m3u, even if the links run via the m3u-filter reverse proxy. There seems to be a difference with files / connection coming from my XT provider. I don't know how to work out what the problem is in this case :-/

**Comment by euzu** [comment]
@jojo141185  i have tested with vlc and your public video list, i could seek in all videos, i had no problems. Could you name the url which makes problems ?

**Comment by jojo141185** [comment]
@euzu Tried to set up a test environment to reproduce the problem. I have provided you access data via e-mail.

**Comment by euzu** [comment]
@jojo141185 did you test whit the latest pushes to the reconnect_stream feature branch ?

**Comment by jojo141185** [comment]
@euzu No, haven't had a chance yet. I'm on a business trip at the moment and won't get around to it until next week.

**Comment by jojo141185** [comment]
@euzu Did you have the opportunity to test with the account I sent you by mail?

**Comment by euzu** [comment]
I want to test but i dont know if you installed the latest version ? If it is the latest version i can test.

**Comment by jojo141185** [comment]
@euzu I have updated m3u-filter based on the current master branch. I have buffering and the retry option disabled. 
The goal would be that you connect via m3u-filter and then activate buffer, test and debug (if it doesn't work)

**Comment by euzu** [comment]
@jojo141185 you have an email.

**Comment by euzu** [comment]
@jojo141185 i could fix the VLC seek problem. Can you please confirm with the latest master?

**Comment by jojo141185** [comment]
I just ran a test with the current master. Seeking now works with the VLC player, even with the buffer and retry options enabled. However, I’m still having issues using Kodi as a player. I don’t understand why every player behaves differently in this regard.
@euzu Could you perhaps take another closer look at this please? Kodi can be installed very easily on Linux via [Flatpak](https://flathub.org/apps/tv.kodi.Kodi).

**Comment by jojo141185** [comment]
@euzu Exactly, I tested it from Kodi with the STRM-file export of m3u-filter.  
I am proceeding as follows:
1. Add Source to Kodi
Go to Settings → Media → Library → Add videos...
Browse to the folder containing the .strm files and set it as a new source
2. Set Content Type
After adding the folder, set the content type (e.g., Movies, TV Shows). 
Choose a scraper for metadata (e.g., TMDb or TVDb) and allow Kodi to scan the folder for updates.
3. Update Library
Kodi scans the .strm folder and adds the items to the library. Metadata is fetched and displayed.
4. Play a Video from library and try seeking by cursor left / right

**Comment by jojo141185** [comment]
I think I have found the problem why seeking is not possible in some players.
When accessing the same content directly or with m3u-filter reverse proxy, two different HTTP headers are returned:

**Direct Access:**
```
< HTTP/1.1 200 OK
< Accept-Ranges: 0-4694430648
< Access-Control-Allow-Origin: *
< Connection: keep-alive
< Content-Length: 4694430648
< Content-Range: bytes 0-4694430647/4694430648
< Content-Type: video/x-matroska
< Date: Thu, 23 Jan 2025 07:39:35 GMT
```

**Access via `m3u-filter` Proxy:**
```
< HTTP/1.1 200 OK
< transfer-encoding: chunked
< date: Thu, 23 Jan 2025 07:44:49 +0000
< accept-ranges: 0-4694430648
< content-range: bytes 0-4694430647/4694430648
< content-type: video/x-matroska
< access-control-allow-credentials: true
< access-control-allow-origin: *
< vary: accept-encoding, Origin, Access-Control-Request-Method, Access-Control-Request-Headers
```

The proxy adds the Transfer-Encoding: chunked header in addition to size-related headers like Content-Length and Content-Range. This violates the HTTP/1.1 specification, which states:

> [RFC 7230, Section 4.1.2:](https://datatracker.ietf.org/doc/html/rfc7230#section-4.1.2)
> A sender MUST NOT generate a trailer that contains a field necessary for message framing (e.g., Transfer-Encoding and Content-Length).

Players like Kodi adheres strictly to this rule, while other players may ignore Transfer-Encoding and rely on Content-Length or Content-Range, allowing seeking despite the misconfiguration.

Kodi disables seeking when the response includes the `Transfer-Encoding: chunked` header. 
This behavior is defined in the Kodi source code [`CurlFile.cpp#L1159`](https://github.com/xbmc/xbmc/blob/f638040be2e1e661a04065343e5c097c172bce65/xbmc/filesystem/CurlFile.cpp#L1159)

```cpp
  if(StringUtils::EqualsNoCase(m_state->m_httpheader.GetValue("Transfer-Encoding"), "chunked"))
    m_state->m_fileSize = 0;

  if(m_state->m_fileSize <= 0)
    m_seekable = false;
```

When Transfer-Encoding: chunked is present, Kodi sets m_fileSize = 0 and disables seeking because it assumes the file size is unknown. This behavior should be the reason for issue https://github.com/euzu/m3u-filter/issues/150.

Accordingly, the reverse proxy in m3u-filter should be adjusted so that the Transfer-Encoding: chunked header is omitted when a Content-Length header is already present, to comply with the HTTP/1.1 specification.

@euzu Can you confirm my assumption and change it accordingly?

**Comment by euzu** [comment]
@jojo141185 i really dont know how you got this headers. I do not add explicit headers except when CONTENT_TYPE, CONNECTION and VARY is missing.

I filter  the provider response and return only 
```
"accept", "content-type", "content-length", "connection", "accept-ranges", "content-range", "vary", "transfer-encoding", "access-control-allow-origin", "access-control-allow-credentials", "icy-metadata"
```

I am a little bit confused why Content-Length is filtered and Transfer-Encoding is added ?

**Comment by euzu** [comment]
@jojo141185 i have added a fix, can you test it with the latest master,  if the same problem occurs i need to debug it to find out the reason.

**Comment by jojo141185** [comment]
@euzu It seems that nothing has changed... :-(
Maybe the headers are added automatically via actix-web? Can you reproduce the problem for yourself?

**Comment by jojo141185** [comment]
@euzu After reviewing your code (I’m an absolute beginner with Rust, so apologies if I’m completely off-track), the issue seems to stem from how the stream response is implemented in actix_web. Currently, it appears that [actix_web::body::BodyStream](https://docs.rs/actix-web/latest/actix_web/body/struct.BodyStream.html) is being used to handle the response. This implementation defaults to using the Transfer-Encoding: chunked header, which is added when the response size is not predefined. However, this conflicts with the Content-Length and Content-Range headers that are also being set in the response. According to the HTTP/1.1 specification (RFC 7230, Section 4.1.2), it is not allowed to include both Transfer-Encoding: chunked and Content-Length headers simultaneously.

Players like Kodi (and probably many others) strictly adhere to this standard and interpret the presence of Transfer-Encoding: chunked as a sign that the file size is unknown. As a result, Kodi disables seeking by setting the file size to 0 internally. This is why seeking works with direct access (where only Content-Length is used) but fails when going through the proxy, as the proxy adds the conflicting Transfer-Encoding: chunked header.

To address this, it may be necessary to distinguish between two scenarios in the code:
- Live streams, where the response size is unknown and BodyStream with Transfer-Encoding: chunked may still be appropriate.
- Videos with a known length, where [actix_web::body::SizedStream ](https://docs.rs/actix-web/latest/actix_web/body/struct.SizedStream.html)should be used instead. This ensures that Content-Length is properly set, and the conflicting Transfer-Encoding: chunked header is omitted, enabling seeking functionality in players like Kodi.

Making this distinction in the routing or response generation logic should probably help to resolve the issue and bring the implementation in line with the HTTP/1.1 specification.

**Comment by euzu** [comment]
@jojo141185 yes you are right. 

What i send and what the client gets are different. I printed he headers at the end of the api handler and saw that conten-length is calculated by actix,  but not send to the client. Instead  chunked header is send.
I manually set the no_chunked flag when there is a Content-Length, but then something goes realy wrong and the Content-Length is a number which does not match the ranges and kodi quits. I really dont know why actix is doing this weird things with the headers.

I am now trying to find out the content-Length and set it manually from the range headers.

**Comment by jojo141185** [comment]
Perfect!
You can extract the size information from the original response CONTENT_LENGTH header:

```
let content_length = provider_response
    .as_ref()
    .and_then(|(_, headers)| headers.get(header::CONTENT_LENGTH))
    .and_then(|val| val.to_str().ok())
    .and_then(|val| val.parse::<u64>().ok());

```
Then, use this value to differentiate between the two Actix Web streaming types:
```
  if let Some(size) = content_length {
      response_builder.body(SizedStream::new(size, stream))
  } else {
      response_builder.body(BodyStream::new(stream))
  }
```

**Comment by euzu** [comment]
@jojo141185 pushed to master,  this made me crazy , i did not expect to have SizedStream for streams with ContentLength, and BodyStream with streams without ContentLength. I expected if there is a ContentLength in the header it is sized.
But we need to do this manually.

**Comment by jojo141185** [comment]
@euzu Looks great! 😊 A quick test shows it's working flawlessly in Kodi — and without any extra props in the STRM file!  🚀
Thanks for implementing it so quickly!

**Comment by euzu** [comment]
@jojo141185 you gave me the right idea. We solved this together. 🎉🎉
